builtin/aws/apprunner/platform.go

```diff
@@ -0,0 +1,445 @@
+	"github.com/aws/aws-sdk-go/service/apprunner"
+	"github.com/aws/aws-sdk-go/service/iam"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/waypoint-plugin-sdk/component"
+	"github.com/hashicorp/waypoint-plugin-sdk/docs"
+	"github.com/hashicorp/waypoint-plugin-sdk/terminal"
+	"github.com/hashicorp/waypoint/builtin/aws/ecr"
+	"github.com/hashicorp/waypoint/builtin/aws/utils"
+)
+
+type Platform struct {
+	config PlatformConfig
+}
+
+type PlatformConfig struct {
+	Region string `hcl:"region,optional"`
+
+	Name string `hcl:"name"`
+
+	Memory int `hcl:"memory,optional"`
+	Cpu    int `hcl:"cpu,optional"`
+
+	// Once created, Port cannot be modified
+	Port int `hcl:"port,optional"`
+	// Once created, RoleName cannot be modified
+	RoleName string `hcl:"role_name,optional"`
+
+	// Environment variables that are meant to configure the application in a static
+	// way. This might be control an image that has multiple modes of operation,
+	// selected via environment variable. Most configuration should use the waypoint
+	// config commands.
+	StaticEnvVars map[string]string `hcl:"static_environment,optional"`
+}
+
+// Config implements Configurable
+func (p *Platform) Config() (interface{}, error) {
+	return &p.config, nil
+}
+
+// DeployFunc implements component.Platform
+func (p *Platform) DeployFunc() interface{} {
+	return p.Deploy
+}
+
+// Deploy creates an AppRunner service, updates an
+// AppRunner service, or no-ops if zero configuration
+// changes are detected.
+func (p *Platform) Deploy(
+	ctx context.Context,
+	log hclog.Logger,
+	src *component.Source,
+	img *ecr.Image,
+	deployConfig *component.DeploymentConfig,
+	ui terminal.UI,
+) (*Deployment, error) {
+
+	sg := ui.StepGroup()
+	defer sg.Wait()
+
+	step := sg.Add("Connecting to AWS")
+	defer func() {
+		step.Abort()
+	}()
+
+	deployment := &Deployment{}
+	if id, err := component.Id(); err != nil {
+		return nil, err
+	} else {
+		deployment.Id = id
+	}
+
+	sess, err := utils.GetSession(&utils.SessionConfig{
+		Region: p.config.Region,
+		Logger: log,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	port := p.config.Port
+	if port == 0 {
+		// App Runner default port
+		port = 8080
+	}
+
+	roleName := p.config.RoleName
+	if roleName == "" {
+		// if role name is not specified, we'll fall back to
+		// getting or creating a service role, with the same
+		// default roleName that AWS uses in the console
+		roleName = defaultRoleName
+	}
+
+	// roleArn only applies to the initial create
+	var roleArn string
+	if arn, err := p.getOrCreateIamRoleArn(sess, log, roleName); err != nil {
+		return nil, err
+	} else {
+		roleArn = arn
+	}
+
+	mem := int64(p.config.Memory)
+	if mem == 0 {
+		mem = defaultMemory
+	}
+
+	cpu := int64(p.config.Cpu)
+	if cpu == 0 {
+		cpu = defaultCpu
+	}
+
+	envVars := make(map[string]*string)
+	for k, v := range p.config.StaticEnvVars {
+		envVars[k] = aws.String(v)
+	}
+
+	step.Done()
+
+	step = sg.Add("Deploying App Runner service %q", p.config.Name)
```

Comment:
Is it necessary to add a new step here, since it's `Done()` in the next line?

Reply:
Hmm I guess not — this mostly stems from my lack of understanding of the exact intended usage of Steps/StepGroups 🤔